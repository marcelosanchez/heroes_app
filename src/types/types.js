export const types = {  // exporto los tipos
    login: '[auth] Login',  // accion de login
    logout: '[auth] Logout',  // accion de logout
}
