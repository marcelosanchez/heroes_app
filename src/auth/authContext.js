import { createContext } from "react";

export const AuthContext = createContext();  // exporto el contexto de autenticacion
