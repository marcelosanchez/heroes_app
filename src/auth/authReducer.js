import { types } from "../types/types";

// const state = {
//     name: 'Marco',
//     logged: true,
// }

// const loginAction = {
//     type: types.login,
//     payload: {
//         name: 'Marco',
//         email: 'marco@mail.com,'
//     }
// }

export const authReducer = ( state = {}, action ) => {  // funcion reductora
    switch ( action.type ) {
        case types.login:  // si el tipo es login
            return {
                ...action.payload,  // retorno el payload con el nombre
                logged: true,
            }
        case types.logout:  // si el tipo es logout
            return {
                logged: false,
            }
        default:
            return state;  // si no hay ninguna accion, devuelvo el estado
    }
};