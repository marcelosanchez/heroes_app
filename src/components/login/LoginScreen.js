import React, { useContext } from 'react'
import { AuthContext } from '../../auth/authContext';
import { useNavigate } from 'react-router-dom';
import { types } from '../../types/types';

export const LoginScreen = () => {

    const navigate = useNavigate();  // permite navegar entre las rutas
    const { dispatch } = useContext(AuthContext);  // !se debe llamar fuera

    const handleLogin = () => {
        const action = {
            type: types.login,
            payload: { name: 'Marcelo' }
        }
        
        dispatch(action);  // se dispara la accion
        
        const lastPath = localStorage.getItem('lastpath') || '/marvel';  // se obtiene la ultima ruta visitada
        navigate(lastPath, { 
            replace: true,  // evita que regrese a la ruta anterior
        })
    };

    return (
        <div className='container mt-5'>
            <h1>Login Screen</h1>
            <hr/>

            <button
                className='btn btn-primary'
                onClick={ handleLogin }
            >
                Login
            </button>
        </div>
    )
}
