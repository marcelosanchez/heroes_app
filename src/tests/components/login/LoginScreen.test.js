import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { AuthContext } from '../../../auth/authContext';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { types } from '../../../types/types';

const mockNavigate = jest.fn();  // mock del navigate

jest.mock('react-router-dom', () => {  // mock del router de react-router-dom
    const original = jest.requireActual('react-router-dom');  // requerimos el original
    return {
        ...original,
        useNavigate: () => mockNavigate,
    };
});

describe('Pruebas en el <LoginScreen />', () => {
    const contextValue = {
        dispatch: jest.fn(),  // mock del dispatch
        user: {
            logged: false,
        },
    };

    const wrapper = mount(
        <AuthContext.Provider value={ contextValue }>
            <MemoryRouter initialEntries={['/login']}>
                <Routes>
                    <Route path="/login" element={<LoginScreen />} />
                </Routes>
            </MemoryRouter>
        </AuthContext.Provider>
    );
    test('Debe de mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe realizar el dispatch y la navegacion', () => {
        const handleClick = wrapper.find('button').prop('onClick');  // se obtiene el onClick del boton que es una funcion
        handleClick();  // se dispara el evento onClick del boton
        
        expect( contextValue.dispatch ).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                name: 'Marcelo',
            }
        });  // se dispara la accion
        
        expect( mockNavigate ).toHaveBeenCalledWith('/marvel', { replace: true });  // replace: true para que no vuelva a la ruta anterior
        
        localStorage.setItem('lastpath', '/dc');  // se guarda la ruta en el localStorage
        
        handleClick();  // se dispara el evento onClick del boton
        
        expect( mockNavigate ).toHaveBeenCalledWith('/dc', { replace: true });  // replace: true para que no vuelva a la ruta anterior

    })
})