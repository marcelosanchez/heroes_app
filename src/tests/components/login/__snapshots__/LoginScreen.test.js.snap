// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <LoginScreen /> Debe de mostrar el componente correctamente 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
